judo/utils/normalization.py

```diff
@@ -0,0 +1,208 @@
+            - 1
+        )
+        return x_normalized
+
+    def denormalize(self, x: np.ndarray) -> np.ndarray:
+        """Denormalize the data."""
+        x_denormalized = x.copy()
+        x_denormalized[..., self.norm_dims] = (x[..., self.norm_dims] + 1) * (
+            self.max[self.norm_dims] - self.min[self.norm_dims]
+        ) / 2 + self.min[self.norm_dims]
+        return x_denormalized
+
+
+class RunningMeanStdNormalizer(Normalizer):
+    """Normalizer that uses running statistics (mean and std) to scale the data."""
+
+    def __init__(
+        self,
+        dim: int,
+        init_std: float = 1.0,
+        min_std: float = 1e-6,
+        max_std: float = 1e6,
+        eps: float = 1e-8,
+    ) -> None:
+        """Initialize the normalizer.
+
+        Args:
+            dim: Dimension of the data.
+            init_std: The initial standard deviation for the running statistics.
+            min_std: The minimum standard deviation.
+            max_std: The maximum standard deviation.
+            eps: Small value to prevent division by zero.
+        """
+        super().__init__(dim)
+        self.eps = eps
+        self.min_std = min_std
+        self.max_std = max_std
+
+        # Running statistics
+        self.count = 0
+        self.mean = np.zeros(dim)
+        self.std = np.ones(dim) * init_std
+
+        # For Welford's online algorithm. M2: sum of squares of differences from the current mean
+        self.M2 = np.zeros(dim)
+
+    def update(self, x: np.ndarray) -> None:
+        """Update the running statistics.
+
+        Args:
+            x: The data to update the running statistics with. Shape=(batch_size, dim).
+        """
+        assert x.ndim == 2, f"Expected 2D array (batch_size, dim), but got {x.ndim}D array {x.shape}"
+        assert x.shape[1] == self.dim, f"Expected dimension {self.dim}, but got {x.shape[1]}"
+
+        batch_size = x.shape[0]
+        self.count += batch_size
+
+        # new values - old mean
+        delta = x - self.mean
+
+        # Update mean
+        self.mean += np.sum(delta, axis=0) / self.count
+
+        # new values - new mean
+        delta2 = x - self.mean
+
+        # Update M2
+        self.M2 += np.sum(delta * delta2, axis=0)
+        self.M2 = np.maximum(self.M2, 0)
+
+        # Update std
+        self.std = np.sqrt(self.M2 / self.count)
+        self.std = np.clip(self.std, self.min_std, self.max_std)
+
+    def normalize(self, x: np.ndarray) -> np.ndarray:
+        """Normalize the data."""
+        return (x - self.mean) / (self.std + self.eps)
+
+    def denormalize(self, x: np.ndarray) -> np.ndarray:
+        """Denormalize the data."""
+        return x * self.std + self.mean
+
+
+normalizer_registry = {
+    "identity": IdentityNormalizer,
+    "min_max": MinMaxNormalizer,
+    "running_mean_std": RunningMeanStdNormalizer,
+}
+
+
+def make_normalizer(normalizer_type: str, dim: int, **kwargs: NormalizerKwargs) -> Normalizer:
```

Comment:
the normalizer type should be a literal. you can declare the valid types earlier in the file so the type isn't super long